cookbooks/cdo-varnish/libraries/helpers.rb

```diff
@@ -120,24 +122,113 @@ def behavior_for_path(behaviors, path)
 
 # Generates the logic string for the specified behavior.
 def process_behavior(behavior, app, section)
-  if section == 'proxy'
-    process_proxy(behavior[:proxy] || app, app)
-  else
-    process_cookies(behavior[:cookies], section)
+  case section
+    when 'proxy'
+      process_proxy(behavior[:proxy] || app, app)
+    when 'vary'
+      process_vary(behavior)
+    when 'request'
+      process_request(behavior)
+    else
+      process_response(behavior)
   end
 end
 
-# Returns the cookie-filter string for a given 'cookies' behavior.
-def process_cookies(cookies, section)
-  if section == 'request'
-    return '# Allow all request cookies.' if cookies == 'all'
-    cookies = ['NO_CACHE'] if cookies == 'none'
-    "cookie.filter_except(\"#{cookies.join(',')}\");"
-  else
-    cookies == 'none' ?
-      'unset beresp.http.set-cookie;' :
-      '# Allow set-cookie responses.'
+# VCL string to create or update a Vary header field with the provided Vary header.
+def set_vary(header, resp)
+  <<VCL
+  if (!#{resp}.http.Vary) {
+    set #{resp}.http.Vary = "#{header}";
+  } elseif (#{resp}.http.Vary !~ "(?<=\\s|,|^)#{header}") {
+    set #{resp}.http.Vary = #{resp}.http.Vary + ", #{header}";
+  }
+VCL
+end
+
+# VCL string to set the appropriate Vary header fields based on the provided cache behavior.
+# Whitelisted headers are added to Vary directly.
+# Whitelisted cookies are added to Vary via their extracted X-COOKIE headers.
```

Comment:
The way this code manages to extract and separately-cache individual cookie key-value pairs is a bit tricky, these are the two main portions:

- Extract the cookies specified in the whitelist from the `Cookie` request header field, transforming them into a set of custom `X-COOKIE-*` HTTP request header fields exposed to the Varnish cache and origin server.
  - (Note: If we didn't want to expose these extra fields to the origin server, I think they could be optionally stripped from the request while still preserving the same cache behavior within Varnish, but I haven't done this extra step yet since I didn't think it was too important.)
- Once the response is returned from the origin, append the custom `X-COOKIE-*` field names to the `Vary` response header field.
- Varnish then automatically includes the extra header fields in its cache-key, and serves the cached response for future requests containing matching header values for those custom `X-COOKIE` fields. (Varnish follows the HTTP/1.1 spec for this behavior, so requires no additional code/configuration.)

So `X-COOKIE` has no magic behavior, I added it solely to support this new functionality. however, I see now that the `X-` prefix has been [deprecated](https://tools.ietf.org/html/rfc6648) as an Internet best practice, so perhaps a more appropriate (if not too verbose) header-field prefix here would be `CDO-VARNISH-COOKIE-*`?